git_services/git_services/init/cloner.py

```diff
@@ -0,0 +1,150 @@
+        self.cli.git_config("push.default simple")
+
+    def _setup_proxy(self, proxy_url):
+        self.cli.git_config(f"http.proxy {proxy_url}")
+        self.cli.git_config("http.sslVerify false")
+
+    @contextmanager
+    def _temp_plaintext_credentials(self):
+        credential_loc = Path("/tmp/git-credentials")
+        with open(credential_loc, "w") as f:
+            f.write(f"https://oauth2:{self.user.oauth_token}@{self.git_host}")
+        yield self.cli.git_config(f"credential.helper 'store --file={credential_loc}'")
+        # NOTE: Temp credentials MUST be cleaned up on context manager exit
+        credential_loc.unlink()
+        self.cli.git_config("--unset credential.helper")
+
+    def _clone(self, branch):
+        lfs_skip_smudge = "" if self.lfs_auto_fetch else "--skip-smudge"
+        self.cli.git_lfs(f"install {lfs_skip_smudge} --local")
+        self.cli.git_remote(f"add {self.remote_name} {self.repo_url}")
+        self.cli.git_fetch(self.remote_name)
+        res = self.cli.git_checkout(branch)
+        _, err = res.communicate()
+        if res.returncode != 0 or len(err) != 0:
+            if b"no space left on device" in err:
+                # INFO: not enough disk space
+                raise errors.NoDiskSpaceError
+            else:
+                # INFO: the branch simply does not exist, create it and continue
+                self.cli.git_checkout(f"-b {branch}")
+        self.cli.git_submodule("init")
+        self.cli.git_submodule("update")
+
+    def _get_autosave_branch(self, session_branch, root_commit_sha):
+        autosave_regex = (
+            f"^{self.remote_origin_prefix}/{self.autosave_branch_prefix}/"
+            f"{self.user.username}/{session_branch}/{root_commit_sha[:7]}/[a-zA-Z0-9]{7}$"
+        )
+        branches = self.cli.git_branch("-a").split()
+        autosave = [
+            branch
+            for branch in branches
+            if re.match(autosave_regex, branch) is not None
+        ]
+        if len(autosave) == 0:
+            return None
+        return autosave[0]
+
+    def _recover_autosave(self, autosave_branch):
+        autosave_items = autosave_branch.split("/")
+        # INFO: Check if the found autosave branch has a valid format, fail otherwise
+        if len(autosave_items) < 7:
+            raise errors.UnexpectedAutosaveFormatError
+        # INFO: Reset the file tree to the auto-saved state.
+        self.cli.git_reset(f"--hard {autosave_branch}")
+        # INFO: Reset HEAD to the last committed change prior to the autosave commit.
+        pre_save_local_commit_sha = autosave_items[7]
+        self.cli.git_reset(f"--soft {pre_save_local_commit_sha}")
+        # INFO: Unstage all modified files.
+        self.cli.git_reset("HEAD .")
+        # INFO: Delete the autosave branch both remotely and locally.
+        autosave_local_branch = autosave_items[2:]
+        self.cli.git_push(f'{self.remote_name} :"{autosave_local_branch}"')
```

Comment:
```suggestion
        self.cli.git_push(f'{self.remote_name} --delete "{autosave_local_branch}"')
```
https://git-scm.com/docs/git-push#Documentation/git-push.txt---delete Does the same but looks leaner/more readable